src/plugins/media/media.js

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    enhanceMediaNodes = function () {
+        var i, id, media, type;
+        
+        //gc = event.detail.api.lib.gc;
+        media = document.querySelectorAll('audio, video');
+        for (i = 0; i < media.length; i++) {
+            type = media[i].nodeName.toLowerCase();
+            // Set an id to identify each media node - used e.g. by the consoleMedia plugin
+            id = media[i].getAttribute('id');
+            if (id === undefined || id === null) {
+                media[i].setAttribute('id', 'media-' + type + '-' + i);
+            }
+            gc.addEventListener(media[i], "play", onPlay);
+            gc.addEventListener(media[i], "playing", onPlay);
+            gc.addEventListener(media[i], "pause", onPause);
+            gc.addEventListener(media[i], "ended", onEnded);
+        }
+    };
+
+    enhanceMedia = function () {
+        var steps, i;
+        enhanceMediaNodes();
+        steps = document.getElementsByClassName('step');
+        for (i = 0; i < steps.length; i++) {
+            gc.addEventListener(steps[i], 'impress:stepenter', onStepenter);
+            gc.addEventListener(steps[i], 'impress:stepleave', onStepleave);
+        }
+    };
+
+    onStepenter = function (event) {
+        var media, play, i;
+        if ((!event) || (!event.target)) {
+            return;
+        }
+        removeMediaClasses();
+        media = event.target.querySelectorAll('audio, video');
+        for (i = 0; i < media.length; i++) {
+            play = media[i].dataset.mediaAutoplay;
+            if (play !== undefined && play !== "false") {
+                media[i].play();
+            }
+        }
+    }
+    
+    onStepleave = function (event) {
+        var media, i, dplay, play, dpause, pause, dstop, stop;
+        if ((!event || !event.target)) {
+            return;
+        }
+        
+        media = event.target.querySelectorAll('audio, video');
+        for (i = 0; i < media.length; i++) {
+            dplay = media[i].dataset.mediaAutoplay;
+            dpause = media[i].dataset.mediaAutopause;
+            dstop = media[i].dataset.mediaAutostop;
+            
+            if (dplay !== undefined && dplay !== "false") {
+                play = true;
+            } else {
+                play = false
+            }
+            
+            if (dpause == undefined || dpause !== "false") {
+                pause = true;
+            } else {
+                pause = false;
+            }
+            
+            if (dstop == undefined) {
+                stop = play && !pause;
+            } else if (dstop !== "false") {
+                stop = true;
+            } else {
+                stop = false;
+            }
+            
+            if (pause || stop) {
+                media[i].pause();
+                if (stop) {
+                    media[i].currentTime = 0;
+                }
+            }
+        }
```

Comment:
I like the general structure here. Previous PR has continued discussion on defaults.